File: app/src/main/java/com/example/myapplication/auth/SignUpActivity.java
package com.example.myapplication.auth;

import android.content.Intent;
import android.os.Bundle;
import android.text.method.HideReturnsTransformationMethod;
import android.text.method.PasswordTransformationMethod;
import android.widget.Button;
import android.widget.EditText;
import android.widget.ImageView;
import android.widget.TextView;
import android.widget.Toast;
import androidx.appcompat.app.AppCompatActivity;
import com.example.myapplication.R;
import com.example.myapplication.models.ApiResponse;
import com.example.myapplication.models.User;
import com.example.myapplication.network.ApiClient;
import java.util.HashMap;
import java.util.Map;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class SignUpActivity extends AppCompatActivity {

    private boolean passVisible = false;
    private boolean confirmPassVisible = false;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sign_up);

        EditText etUsername = findViewById(R.id.et_username);
        EditText etEmail = findViewById(R.id.et_email);
        EditText etPassword = findViewById(R.id.et_password);
        EditText etConfirmPassword = findViewById(R.id.et_confirm_password);
        Button btnSignup = findViewById(R.id.btn_signup);
        ImageView ivTogglePass = findViewById(R.id.iv_toggle_pass);
        ImageView ivToggleConfirm = findViewById(R.id.iv_toggle_confirm);
        TextView tvLoginLink = findViewById(R.id.tv_login_link);

        // Toggle password
        ivTogglePass.setOnClickListener(v -> {
            if (passVisible) {
                etPassword.setTransformationMethod(PasswordTransformationMethod.getInstance());
                ivTogglePass.setImageResource(R.drawable.ic_visibility);
                passVisible = false;
            } else {
                etPassword.setTransformationMethod(HideReturnsTransformationMethod.getInstance());
                ivTogglePass.setImageResource(R.drawable.ic_visibility_off);
                passVisible = true;
            }
            etPassword.setSelection(etPassword.getText().length());
        });

        // Toggle confirm password
        ivToggleConfirm.setOnClickListener(v -> {
            if (confirmPassVisible) {
                etConfirmPassword.setTransformationMethod(PasswordTransformationMethod.getInstance());
                ivToggleConfirm.setImageResource(R.drawable.ic_visibility);
                confirmPassVisible = false;
            } else {
                etConfirmPassword.setTransformationMethod(HideReturnsTransformationMethod.getInstance());
                ivToggleConfirm.setImageResource(R.drawable.ic_visibility_off);
                confirmPassVisible = true;
            }
            etConfirmPassword.setSelection(etConfirmPassword.getText().length());
        });

        // Tombol Sign Up
        btnSignup.setOnClickListener(v -> {
            String username = etUsername.getText().toString().trim();
            String email = etEmail.getText().toString().trim();
            String password = etPassword.getText().toString().trim();
            String confirmPassword = etConfirmPassword.getText().toString().trim();

            if (username.isEmpty() || email.isEmpty() || password.isEmpty() || confirmPassword.isEmpty()) {
                Toast.makeText(this, "Semua field harus diisi", Toast.LENGTH_SHORT).show();
                return;
            }

            if (!password.equals(confirmPassword)) {
                Toast.makeText(this, "Password tidak cocok", Toast.LENGTH_SHORT).show();
                return;
            }

            if (password.length() < 6) {
                Toast.makeText(this, "Password minimal 6 karakter", Toast.LENGTH_SHORT).show();
                return;
            }

            Map<String, String> body = new HashMap<>();
            body.put("username", username);
            body.put("email", email);
            body.put("password", password);

            ApiClient.getApiService().register(body).enqueue(new Callback<ApiResponse<User>>() {
                @Override
                public void onResponse(Call<ApiResponse<User>> call,
                                       Response<ApiResponse<User>> response) {
                    if (response.isSuccessful() && response.body() != null
                            && response.body().isSuccess()) {
                        Toast.makeText(SignUpActivity.this,
                                "Akun berhasil dibuat!", Toast.LENGTH_SHORT).show();
                        startActivity(new Intent(SignUpActivity.this, LoginActivity.class));
                        finish();
                    } else {
                        Toast.makeText(SignUpActivity.this,
                                "Gagal membuat akun", Toast.LENGTH_SHORT).show();
                    }
                }

                @Override
                public void onFailure(Call<ApiResponse<User>> call, Throwable t) {
                    Toast.makeText(SignUpActivity.this,
                            "Tidak dapat terhubung ke server", Toast.LENGTH_SHORT).show();
                }
            });
        });

        // Link ke Login
        tvLoginLink.setOnClickListener(v -> {
            startActivity(new Intent(this, LoginActivity.class));
            finish();
        });
    }
}

File: app/src/main/java/com/example/myapplication/auth/SignUpActivity.kt
package com.example.myapplication.auth

import android.content.Intent
import android.os.Bundle
import android.text.method.HideReturnsTransformationMethod
import android.text.method.PasswordTransformationMethod
import android.widget.Button
import android.widget.EditText
import android.widget.ImageView
import android.widget.TextView
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.lifecycleScope
import com.example.myapplication.R
import com.example.myapplication.network.SupabaseClientProvider
import io.github.jan.supabase.auth.auth
import io.github.jan.supabase.auth.providers.builtin.Email
import io.github.jan.supabase.postgrest.postgrest
import kotlinx.coroutines.launch
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.put
import kotlinx.serialization.json.JsonObject
import kotlinx.serialization.json.JsonPrimitive

class SignUpActivity : AppCompatActivity() {

    private var passVisible = false
    private var confirmPassVisible = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sign_up)

        val etUsername: EditText = findViewById(R.id.et_username)
        val etEmail: EditText = findViewById(R.id.et_email)
        val etPassword: EditText = findViewById(R.id.et_password)
        val etConfirm: EditText = findViewById(R.id.et_confirm_password)
        val btnSignup: Button = findViewById(R.id.btn_signup)
        val ivTogglePass: ImageView = findViewById(R.id.iv_toggle_pass)
        val ivToggleConfirm: ImageView = findViewById(R.id.iv_toggle_confirm)
        val tvLogin: TextView = findViewById(R.id.tv_login_link)

        ivTogglePass.setOnClickListener {
            passVisible = !passVisible
            etPassword.transformationMethod = if (passVisible)
                HideReturnsTransformationMethod.getInstance()
            else
                PasswordTransformationMethod.getInstance()
            ivTogglePass.setImageResource(
                if (passVisible) R.drawable.ic_visibility_off else R.drawable.ic_visibility
            )
            etPassword.setSelection(etPassword.text.length)
        }

        ivToggleConfirm.setOnClickListener {
            confirmPassVisible = !confirmPassVisible
            etConfirm.transformationMethod = if (confirmPassVisible)
                HideReturnsTransformationMethod.getInstance()
            else
                PasswordTransformationMethod.getInstance()
            ivToggleConfirm.setImageResource(
                if (confirmPassVisible) R.drawable.ic_visibility_off else R.drawable.ic_visibility
            )
            etConfirm.setSelection(etConfirm.text.length)
        }

        btnSignup.setOnClickListener {
            val username = etUsername.text.toString().trim()
            val email = etEmail.text.toString().trim()
            val password = etPassword.text.toString().trim()
            val confirm = etConfirm.text.toString().trim()

            if (username.isEmpty() || email.isEmpty() ||
                password.isEmpty() || confirm.isEmpty()) {
                Toast.makeText(this, "Semua field harus diisi", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            if (password != confirm) {
                Toast.makeText(this, "Password tidak cocok", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            if (password.length < 6) {
                Toast.makeText(this, "Password minimal 6 karakter", Toast.LENGTH_SHORT).show()
                return@setOnClickListener
            }

            lifecycleScope.launch {
                try {
                    // Step 1: Register ke Supabase Auth
                    SupabaseClientProvider.client.auth.signUpWith(Email) {
                        this.email = email
                        this.password = password
                        data = buildJsonObject {
                            put("username", username)
                        }
                    }

                    // Step 2: Insert manual ke profiles sebagai fallback
                    try {
                        val userId = SupabaseClientProvider.client
                            .auth.currentSessionOrNull()?.user?.id

                        if (userId != null) {
                            val profileData = JsonObject(
                                mapOf(
                                    "id" to JsonPrimitive(userId),
                                    "email" to JsonPrimitive(email),
                                    "username" to JsonPrimitive(username)
                                )
                            )
                            SupabaseClientProvider.client
                                .postgrest["profiles"]
                                .insert(profileData)
                        }
                    } catch (_: Exception) {
                        // Abaikan jika trigger sudah insert duluan
                    }

                    Toast.makeText(
                        this@SignUpActivity,
                        "Akun berhasil dibuat!",
                        Toast.LENGTH_SHORT
                    ).show()

                    startActivity(
                        Intent(this@SignUpActivity, LoginActivity::class.java)
                    )
                    finish()

                } catch (e: Exception) {
                    Toast.makeText(
                        this@SignUpActivity,
                        "Gagal: ${e.message}",
                        Toast.LENGTH_LONG
                    ).show()
                }
            }
        }

        tvLogin.setOnClickListener {
            startActivity(Intent(this, LoginActivity::class.java))
            finish()
        }
    }
}